### PR TITLE
Add machine deployment for gke

### LINKED ISSUE #4683: Add create machine deployment to GKE cluster
providing an option to create Machine Deployment for GKE external cluster

### PR DESCRIPTION
**Pull request by ahmadhamzh**
**What does this PR do / Why do we need it**:
add create Machine Deployment for GKE External Cluster

**Which issue(s) this PR fixes** :

Fixes #4683

```release-note
Add support to create machine deployment for GKE external cluster.
```

### CODE REVIEW

File: src/app/core/services/external-cluster.ts [hunk 1]
```diff
@@ -234,6 +235,21 @@ export class ExternalClusterService {
     return this._http.get<GKEZone[]>(url, {headers: this._getGKEHeaders()}).pipe(catchError(() => of<[]>()));
   }
 
+  getGKEDiskTypes(projectID: string, clusterID: string): Observable<GCPDiskType[]> {
```
Comment (KhizerRehan) [inline src/app/core/services/external-cluster.ts:238]:
```suggestion
  getGKEDiskTypesForMachineDeployment(projectID: string, clusterID: string): Observable<GCPDiskType[]> {
```

- update all methods which are related to `MD` creation
- Inlcuding EKS one to remove `Create` to be consistent in all providers

File: src/app/external-cluster-wizard/steps/external-cluster/provider/aks/template.html [hunk 2]
```diff
@@ -116,7 +116,7 @@
         Node Pool Name is <strong>required</strong>.
       </mat-error>
       <mat-error *ngIf="form.get(Controls.NodePoolName).hasError(ErrorType.Pattern)">
-        Node Pool Name must contain only lowercase letters and numbers, must be 12 or fewer characters.</mat-error>
+        Node pool name must contain only lowercase letters and numbers, must be 12 or fewer characters.</mat-error>
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/aks/template.html:119]: revert this

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts [hunk 3]
```diff
@@ -57,12 +79,21 @@ export class GKEClusterSettingsComponent
   implements OnInit, OnDestroy, ControlValueAccessor, Validator
 {
   readonly Controls = Controls;
+  readonly AUTOSCALING_MIN_VALUE = 1;
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts:82]: used as `const` inside [here](https://github.com/kubermatic/dashboard/pull/4719/files#diff-1f995b8e4709882957734ef0745d042f9593c9ac126b720983a7d9bf8f171f06R160)

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts [hunk 4]
```diff
@@ -71,7 +102,7 @@ export class GKEClusterSettingsComponent
   ngOnInit(): void {
     this._initForm();
     this._initSubscriptions();
-    this._getGKEZones();
+    !this.isDialogView() && this._getGKEZones();
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts:105]: Avoid single line shortcuts
```suggestion
 if(!this.isDialogView()) {
      this._getGKEZones();
    }
```

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts [hunk 5]
```diff
@@ -84,20 +115,58 @@ export class GKEClusterSettingsComponent
...
+      this.control(Controls.MinCount).updateValueAndValidity();
+    }
+  }
+
   private _initForm(): void {
+    const MIN_COUNT_DEFAULT_VALUE = 1;
+    const MAX_COUNT_DEFAULT_VALUE = 5;
     this.form = this._builder.group({
-      [Controls.Name]: this._builder.control('', [Validators.required, KUBERNETES_RESOURCE_NAME_PATTERN_VALIDATOR]),
+      [Controls.Name]: this._builder.control('', [Validators.required, GKE_POOL_NAME_VALIDATOR]),
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts:138]: Do we need same `regex` in Cluster Creation one for GKE?

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 6]
```diff
@@ -47,7 +49,8 @@
         Name is <strong>required</strong>.
       </mat-error>
       <mat-error *ngIf="form.get(Controls.Name).hasError('pattern')">
-        Name can only contain alphanumeric characters and dashes (a-z, 0-9 and -). Must not start/end with dash.
+        Node pool names must start with a lowercase letter followed by up to 39 lowercase letters,
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:52]: fix padding when with `below` once this error is generated.

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 7]
```diff
@@ -63,20 +66,72 @@
...
+    </km-number-stepper>
+
+    <mat-checkbox [formControlName]="Controls.EnableAutoScaling"
+                  (change)="onEnableAutoScalingChange($event)">
+      Enable Auto Scaling
+      <i class="km-icon-info km-pointer"
+         matTooltip="Indicates whether to enable auto scaling.This option is recommended so that cluster sized correctly for the current running workloads."></i>
+    </mat-checkbox>
+
+    <ng-container *ngIf="form.get(Controls.EnableAutoScaling).value">
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:112]: Check **AKS** Cluster wizard

OR do we need this Create wizard

move right side

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 8]
```diff
@@ -63,20 +66,72 @@
...
-      <mat-error *ngIf="form.get(Controls.NodeCount).hasError('required')">
-        Initial Node Count is <strong>required</strong>.
-      </mat-error>
-    </mat-form-field>
+    <km-autocomplete *ngIf="isDialogView()"
+                     label="Disk Types"
+                     [options]="diskTypes"
+                     [formControlName]="Controls.DiskTypes">
+      <mg-container hint>
+        Set Disk Type
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:74]: - Update `mat-hints` for ALL.

**Review by KhizerRehan — CHANGES_REQUESTED** [review]
Added feedback

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 9]
```diff
@@ -63,20 +66,72 @@
...
+    <km-autocomplete *ngIf="isDialogView()"
+                     label="Machine Sizes"
+                     [options]="machineSizes"
+                     [formControlName]="Controls.MachineSizes">
+      <mg-container hint>
+        Set Machine Size
+      </mg-container>
+    </km-autocomplete>
+
+    <km-autocomplete *ngIf="isDialogView()"
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:87]: - Remove this as per discussion and due to error

`googleapi: Error 400: Creation of node pools using node images based on Docker container runtimes is not supported in GKE v1.24+ clusters as Dockershim has been removed in Kubernetes v1.24., badRequest`

- API call too

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 10]
```diff
@@ -63,20 +66,72 @@
...
+                     label="Disk Types"
+                     [options]="diskTypes"
+                     [formControlName]="Controls.DiskTypes">
+      <mg-container hint>
+        Set Disk Type
+      </mg-container>
+    </km-autocomplete>
+
+    <km-autocomplete *ngIf="isDialogView()"
+                     label="Machine Sizes"
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:79]:
```suggestion
                     label="Machine Size"
```

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 11]
```diff
@@ -63,20 +66,72 @@
...
-      <input matInput
-             type="number"
-             [formControlName]="Controls.NodeCount" />
-      <mat-hint>Set the Initial Node Count.</mat-hint>
-      <mat-error *ngIf="form.get(Controls.NodeCount).hasError('required')">
-        Initial Node Count is <strong>required</strong>.
-      </mat-error>
-    </mat-form-field>
+    <km-autocomplete *ngIf="isDialogView()"
+                     label="Disk Types"
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:70]:
```suggestion
                     label="Disk Type"
```

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html [hunk 12]
```diff
@@ -63,20 +66,72 @@
...
+                     [options]="diskTypes"
+                     [formControlName]="Controls.DiskTypes">
+      <mg-container hint>
+        Set Disk Type
+      </mg-container>
+    </km-autocomplete>
+
+    <km-autocomplete *ngIf="isDialogView()"
+                     label="Machine Sizes"
+                     [options]="machineSizes"
```
Comment (KhizerRehan) [inline src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:80]: show `name + description` you need to update options in `ts` file using `map` operator

**Review by KhizerRehan — CHANGES_REQUESTED** [review]
Added more feedback

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts (hunk 5 above)
Comment (ahmadhamzh) [reply src/app/external-cluster-wizard/steps/external-cluster/provider/gke/component.ts:138]: yes 
Screenshot from 2022-07-28 13-03-34

File: src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html (hunk 7 above)
Comment (ahmadhamzh) [reply src/app/external-cluster-wizard/steps/external-cluster/provider/gke/template.html:112]: no we don't need it here

**Pull request merged**